Implement chained timing functions

steveblock opened the pull request.

steveblock committed:
Add LinearTimingFunction

Currently we implement a linear timing function by setting
Timing._timingFunction to null and special-casing this in TimedItem. Adding a
separate type will simplify the logic in TimedItem and will greatly simplify
chained timing functions.

birtles commented:
Wow, that was fast :)

One thing I just noticed though is that from a quick skim of the test cases this seems to allow specifying easingTimes as a space-separated string but currently the API only lets you specify "align" "distribute" or a sequence of numbers.

steveblock commented:
> Wow, that was fast :)

It was just good timing - I'd just started when you sent out the spec update!

> currently the API only lets you specify "align" "distribute" or a sequence of numbers.

Ah, I hadn't even noticed that. What's the reason for the restriction? All the required logic is already there.

mithro commented:
Test currently fail....

dstoc commented on test/testcases/unit-test-chained-timing-function.html:
```diff
@@ -0,0 +1,239 @@
+<!--
```
This should be impl-test since it uses _WebAnimationsTestingUtilities

birtles commented:
> > currently the API only lets you specify "align" "distribute" or a sequence of numbers.
> 
> Ah, I hadn't even noticed that. What's the reason for the restriction? All the required logic is already there.

The restriction is that you can't specify a sequence of numbers as a string. The reason is that working with numbers is easier than having to parse the string and put it back together again. And we don't support both because it would complicate the code for the author when they go to read back values, e.g.: test if string or sequence, if string test if keyword value or space-separated list of numbers, if sequence of numbers parse etc. etc.

dstoc commented on test/testcases/test-chained-timing-function.html:
```diff
@@ -0,0 +1,77 @@
+<!--
```
Should also add a few test cases to auto-test-timing-functions.html

dstoc commented on web-animations.js:
```diff
@@ -2248,7 +2242,122 @@ TimingFunction.createFromString = function(spec, timedItem) {
+  var result = [];
+  var lastTimingFunction;
+  for (var i = 1; i < points.length; i++) {
+    var range = { min: points[i - 1], max: points[i] };
```
no spaces inside single line object initializer {foo: bar}

dstoc commented:
lgtm

steveblock committed:
Implement chained timing functions

These are specified through the following attributes on Timing ...
- easing: A space-separated list of timing functions
- easingPoints: An array of the end points of the input/output ranges over
  which the timing functions apply. The following keywords are also
  allowed ...
  - distribute: Distribute timing functions evenly in time.
  - align: If the TimedItem with which this Timing object is associated is an
    Animation with a keyframe or path effect, align timing functions with
    keyframes.

See http://dev.w3.org/fxtf/web-animations/#timing-function-chains for details.

Note that alignment with a path effect is not yet implemented.

steveblock commented:
Updated to fix style, add more tests and fix a couple of bugs that were breaking existing tests.

steveblock commented:
> The restriction is that you can't specify a sequence of numbers as a string.

Got it: I misunderstood your first comment. I've updated the logic to require the list of points to be an array of numbers.

Not yet for landing, as Shane has also started an implementation and we're going to pick the best bits of both.

steveblock commented:
I've talked to Shane and I'm going to merge aspects of our implementations. In the mean time, we can land the preparatory commits, which I've moved to https://github.com/web-animations/web-animations-js/pull/452

steveblock commented:
Updated to use the best bits of Shane's and my implementations, fixed a few more bugs (from both) and added more tests.

dstoc commented on web-animations.js:
```diff
@@ -2305,39 +2326,200 @@ var TimingFunction = function() {
+      max: normalizedPositionList[i]
+    };
+    var timingFunction = i - 1 < chain.length ?
+        chain[i - 1] : lastTimingFunction.clone();
```
Why do we need to use clone() here?

dstoc replied:
Oh, because PacedTimingFunction is based on the range.

steveblock replied:
Right, see the last test in impl-test-chained-timing-function.html

dstoc commented on web-animations.js:
```diff
@@ -2305,39 +2326,200 @@ var TimingFunction = function() {
+  while (typeof easing === 'string') {
+    easing = easing.trim();
+    var result = TimingFunction.createComponentFromString(easing, timedItem);
+    if (result === null) {
```
Does this happen when we can't parse a timing function? Shouldn't we fall back to linear in this case?

steveblock replied:
Good catch Fixed.

dstoc commented:
lgtm

steveblock committed:
Use a single linear timing function if the specified string is invalid
